avocado/utils/asset.py

```diff
@@ -71,21 +71,178 @@ def __init__(self, name, asset_hash, algorithm, locations, cache_dirs,
         :param expire: time in seconds for the asset to expire
         :param metadata: metadata which will be saved inside metadata file
         """
-        self.name = name
+        parsed_url = urlparse.urlparse(name)
+        # Asset file name
+        self.basename = os.path.basename(parsed_url.path)
         self.asset_hash = asset_hash
-        if algorithm is None:
-            self.algorithm = DEFAULT_HASH_ALGORITHM
-        else:
+
+        if algorithm:
             self.algorithm = algorithm
+        else:
+            self.algorithm = DEFAULT_HASH_ALGORITHM
 
+        self.locations = []
+        # If `name` is an url, move it to `locations`
+        if parsed_url.scheme:
+            self.locations.append(os.path.dirname(parsed_url.geturl()))
+        # Let's be conservative and check `locations` type
         if isinstance(locations, str):
-            self.locations = [locations]
+            # remove asset name from location
+            if self.basename in locations:
+                parsed_locations = urlparse.urlparse(locations)
+                self.locations.append(
+                    os.path.dirname(parsed_locations.geturl()))
+            else:
+                self.locations.append(locations)
+        elif isinstance(locations, list):
+            # remove asset name from location
+            for location in locations:
+                if self.basename in location:
+                    parsed_location = urlparse.urlparse(location)
+                    self.locations.append(
+                        os.path.dirname(parsed_location.geturl()))
+                else:
+                    self.locations.append(location)
+
+        # Let's make the user's life easy and handle the cache
+        if cache_dirs:
+            self.cache_dirs = cache_dirs
         else:
-            self.locations = locations
-        self.cache_dirs = cache_dirs
+            # Break early if we don't have the cache_dirs. This is necessary
+            # so we don't change the Asset signature.
+            raise EnvironmentError("At least one cache directory is necessary"
+                                   " to download the asset.")
+
         self.expire = expire
         self.metadata = metadata
 
+        # Break early. We need to know the source location of the asset.
+        if not self.locations:
```

Comment:
Would make sense move this block up when it finished all calculations to locations?

Reply:
Absolutely!